dev/lib/product_taxonomy/models/category.rb

```diff
@@ -2,15 +2,246 @@
+    validate :id_starts_with_parent_id, unless: :root?
+    validate :attributes_found?
+    validate :children_found?
+    validate :secondary_children_found?
+    validates_with ProductTaxonomy::ModelIndex::UniquenessValidator, attributes: [:id]
+
+    attr_reader :id, :name, :children, :secondary_children, :attributes, :uniqueness_context
+    attr_accessor :parent, :secondary_parents
+
+    # @param id [String] The ID of the category.
+    # @param name [String] The name of the category.
+    # @param attributes [Array<Attribute>] The attributes of the category.
+    # @param uniqueness_context [ModelIndex] The uniqueness context for the category.
+    # @param parent [Category] The parent category of the category.
+    def initialize(id:, name:, attributes: [], uniqueness_context: nil, parent: nil)
       @id = id
       @name = name
       @children = []
+      @secondary_children = []
       @attributes = attributes
-      @parent = nil
+      @parent = parent
+      @secondary_parents = []
+      @uniqueness_context = uniqueness_context
+    end
+
+    #
+    # Manipulation
+    #
+
+    # Add a child to the category
+    #
+    # @param [Category|String] child node, or the friendly ID if the node was not found.
+    def add_child(child)
+      @children << child
+
+      return unless child.is_a?(Category)
+
+      child.parent = self
+    end
+
+    # Add a secondary child to the category
+    #
+    # @param [Category|String] child node, or the friendly ID if the node was not found.
+    def add_secondary_child(child)
+      @secondary_children << child
+
+      return unless child.is_a?(Category)
+
+      child.secondary_parents << self
+    end
+
+    #
+    # Information
+    #
+    def inspect
+      "#<#{self.class.name} id=#{id} name=#{name}>"
+    end
+
+    # Whether the category is the root category
+    #
+    # @return [Boolean]
+    def root?
+      parent.nil?
+    end
+
+    # Whether the category is a leaf category
+    #
+    # @return [Boolean]
+    def leaf?
+      children.empty?
+    end
+
+    # The level of the category
+    #
+    # @return [Integer]
+    def level
+      ancestors.size
+    end
+
+    # The root category in this category's tree
+    #
+    # @return [Category]
+    def root
+      ancestors.last || self
+    end
+
+    # The ancestors of the category
+    #
+    # @return [Array<Category>]
+    def ancestors
+      if root?
+        []
+      else
+        [parent] + parent.ancestors
+      end
+    end
+
+    # The full name of the category
+    #
+    # @return [String]
+    def full_name
+      return name if root?
+
+      parent.full_name + " > " + name
+    end
+
+    # Whether the category is a descendant of another category
+    #
+    # @param [Category] category
+    # @return [Boolean]
+    def descendant_of?(category)
+      ancestors.include?(category)
+    end
+
+    # Iterate over the category and all its descendants
+    #
+    # @yield [Category]
+    def traverse(&block)
```

Comment:
I assume usage of this will come in later PRs?

Reply:
For now I've been using it for debugging, ex. for recursively printing out the taxonomy